custom/abt/reports/late_pmt.py

```diff
@@ -139,66 +80,85 @@ def headers(self):
+        ))
+        return {
+            (case.received_on, case.operation_site)
+            for case in get_cases_from_forms(self.domain, forms)
+        }
 
     @property
     def rows(self):
-        def _to_report_format(date, user, error_msg):
+        def _to_report_format(date, location, error_msg):
             return [
                 date.strftime("%Y-%m-%d"),
-                user['username'].split('@')[0],
-                user['phone_number'],
-                user['country'],
-                user['level_1'],
-                user['level_2'],
-                user['level_3'],
-                user['level_4'],
+                location.name,
+                location.country,
+                location.level_1,
+                location.level_2,
+                location.level_3,
+                location.level_4,
                 error_msg
             ]
 
-        users = self.get_users
+        include_missing_pmt_data = self.report_config['submission_status'] != 'group_b'
+        # include_incorrect_pmt_data is no longer applicable, because an
+        # invalid SMS will not result in a pmt_data case being created/updated
+        include_incorrect_pmt_data = self.report_config['submission_status'] != 'group_a'
+        error_msg = _('Incorrect or no PMT data submitted')
         dates = rrule(
             DAILY,
             dtstart=self.startdate,
             until=self.enddate,
             byweekday=(MO, TU, WE, TH, FR, SA)
         )
-        include_missing_pmt_data = self.report_config['submission_status'] != 'group_b'
-        include_incorrect_pmt_data = self.report_config['submission_status'] != 'group_a'
         rows = []
-        if users:
-            for date in dates:
-                for user in users:
-                    sms_received = (date.date(), user['user_id']) in self.smss_received
-                    valid_sms = (date.date(), user['user_id']) in self.valid_smss_received
-                    if not sms_received and include_missing_pmt_data:
-                        error_msg = _('No PMT data Submitted')
-                    elif sms_received and not valid_sms and include_incorrect_pmt_data:
-                        error_msg = _('Incorrect PMT data Submitted')
-                    else:
-                        continue
-                    rows.append(_to_report_format(date, user, error_msg))
+        for date in dates:
+            for location in get_locations(self.domain, self.report_config):
+                pmt_submitted = (date.date(), location.id) in self.pmts_submitted
+                if not pmt_submitted and (include_missing_pmt_data or include_incorrect_pmt_data):
+                    rows.append(_to_report_format(date, location, error_msg))
         return rows
+
+
+def iter_forms_by_xmlns_received_on(
+    domain: str,
+    xmlns: str,
+    start_datetime: datetime,
+    end_datetime: datetime,
+) -> Iterator[XFormInstanceSQL]:
```

Comment:
I'm kind of surprised this doesn't exist already - did you consider putting it in the xform accessors?

Reply:
(I was also surprised it wasn't already there.) The reason I chose not to was that at the moment it's only used once, in custom code. If you think we would use it if it was there, I could easily be persuaded to move it.